### PR TITLE
Added documentation and lint checks to PlaceBeacon

### PR DESCRIPTION
**Pull request by Mailaender**
Split from https://github.com/OpenRA/OpenRA/pull/10756.

### CODE REVIEW

**Comment by abcdefg30** [comment]
Looks good to me. :+1: / :white_check_mark:

File: OpenRA.Mods.Common/Effects/Beacon.cs [hunk 1]
```diff
@@ -91,7 +105,7 @@ public IEnumerable<IRenderable> Render(WorldRenderer r)
 			if (!owner.IsAlliedWith(owner.World.RenderPlayer))
 				yield break;
 
-			var palette = r.Palette(palettePrefix + owner.InternalName);
+			var palette = r.Palette(isPlayerPalette ? beaconPalette + owner.InternalName : beaconPalette);
 			foreach (var a in circles.Render(position, palette))
```
Comment (obrakmann) [inline OpenRA.Mods.Common/Effects/Beacon.cs]: This line and the equivalent for `arrow` below will need a null check as well.

File: OpenRA.Mods.Common/Scripting/Global/BeaconGlobal.cs [hunk 2]
```diff
@@ -27,9 +27,9 @@ public BeaconGlobal(ScriptContext context) : base(context)
 
 		[Desc("Creates a new beacon that stays for the specified time at the specified WPos. " +
 			"Does not remove player set beacons, nor gets removed by placing them.")]
-		public Beacon New(Player owner, WPos position, int duration = 30 * 25, bool showRadarPings = true, string palettePrefix = "player")
+		public Beacon New(Player owner, WPos position, int duration = 30 * 25, bool showRadarPings = true)
 		{
-			var playerBeacon = new Beacon(owner, position, duration, palettePrefix);
+			var playerBeacon = new Beacon(owner, position, duration, "player", true, "beacon", "arrow", "circle");
```
Comment (obrakmann) [inline OpenRA.Mods.Common/Scripting/Global/BeaconGlobal.cs]: I'm not sure if that's the best way. How about adding default arguments instead?

**Comment by Mailaender** [comment]
I removed https://github.com/OpenRA/OpenRA/pull/10918/commits/777c1dfe8e22c1f461284b5ebf86254b2a912048 now that I went for a second simplified trait instead of making everything optional.

File: OpenRA.Mods.Common/Scripting/Global/BeaconGlobal.cs (hunk 2 above)
Comment (Mailaender) [inline OpenRA.Mods.Common/Scripting/Global/BeaconGlobal.cs]: That is just shifting hardcoding around. Moved it to a trait lookup instead.

File: OpenRA.Mods.Common/Traits/Player/PlaceBeacon.cs [hunk 3]
```diff
@@ -14,12 +14,19 @@
...
 	{
 		public readonly int Duration = 30 * 25;
 		public readonly string NotificationType = "Sounds";
 		public readonly string Notification = "Beacon";
-		public readonly string PalettePrefix = "player";
+
+		public readonly bool IsPlayerPalette = true;
+		[PaletteReference("IsPlayerPalette")] public readonly string Palette = "player";
+
+		public readonly string BeaconCollection = "beacon";
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Player/PlaceBeacon.cs]: The `Collection` terminology is used for UI artwork, and is incorrect for traits.  See the discussion in #10938.

**Comment by Mailaender** [comment]
Went through rebase hell to make the field names more consistent.

File: OpenRA.Mods.Common/Traits/SupportPowers/SupportPower.cs [hunk 4]
```diff
@@ -43,11 +43,18 @@ public abstract class SupportPowerInfo : UpgradableTraitInfo
 		[Desc("Palette used for the icon.")]
 		[PaletteReference] public readonly string IconPalette = "chrome";
 
-		[Desc("Beacons are only supported on the Airstrike and Nuke powers")]
+		[Desc("Beacons are only supported on the Airstrike, Paratroopers and Nuke powers")]
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/SupportPowers/SupportPower.cs]: Comma after Paratroopers please.

Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/SupportPowers/SupportPower.cs]: Added.

Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/SupportPowers/SupportPower.cs]: For the extra style.  :tophat:

**Comment by Phrohdoh** [comment]
This all looks good to me, but I haven't tested it yet.
I will give this a proper +1 after testing.

File: OpenRA.Mods.Common/Traits/SupportPowers/SupportPower.cs [hunk 5]
```diff
@@ -45,9 +45,11 @@ public abstract class SupportPowerInfo : UpgradableTraitInfo
 
 		[Desc("Beacons are only supported on the Airstrike and Nuke powers")]
 		public readonly bool DisplayBeacon = false;
-		public readonly string BeaconPalettePrefix = "player";
-		public readonly string BeaconPoster = null;
+		[PaletteReference(true)] public readonly string BeaconPalettePrefix = "player";
+		public readonly string BeaconImage = "beacon";
+		[SequenceReference("BeaconImage")] public readonly string BeaconPoster = null;
 		[PaletteReference] public readonly string BeaconPosterPalette = "chrome";
+		[SequenceReference("BeaconImage")] public readonly string ClockSprite = "clock";
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/SupportPowers/SupportPower.cs]: I think this should be `ClockSequence`.

Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/SupportPowers/SupportPower.cs]: Renamed, rebased and solved all conflicts.

File: OpenRA.Mods.Common/Traits/SupportPowers/SupportPower.cs [hunk 6]
```diff
@@ -50,6 +50,9 @@ public abstract class SupportPowerInfo : UpgradableTraitInfo
 		[SequenceReference("BeaconImage")] public readonly string BeaconPoster = null;
 		[PaletteReference] public readonly string BeaconPosterPalette = "chrome";
 		[SequenceReference("BeaconImage")] public readonly string ClockSequence = "clock";
+		[SequenceReference("BeaconImage")] public readonly string ClockSprite = "clock";
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/SupportPowers/SupportPower.cs]: This looks like a rebase error, look at the line above.

File: OpenRA.Mods.Common/Traits/Player/PlaceBeacon.cs [hunk 7]
```diff
@@ -19,8 +19,14 @@ public class PlaceBeaconInfo : ITraitInfo
 		public readonly int Duration = 30 * 25;
 		public readonly string NotificationType = "Sounds";
 		public readonly string Notification = "Beacon";
+
+		[PaletteReference(true)]
 		public readonly string PalettePrefix = "player";
 
+		public readonly string BeaconImage = "beacon";
+		[SequenceReference("BeaconImage")] public readonly string ArrowSprite = "arrow";
+		[SequenceReference("BeaconImage")] public readonly string CircleSprite = "circles";
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Player/PlaceBeacon.cs]: Shouldn't these be `ArrowSequence` and `CircleSequence` now? You call them that in the support power traits already.

**Comment by reaperrr** [comment]
Umm, github thinks the diffs are outdated, but I think they're not, so my comments should be valid.

**Comment by pchote** [comment]
> Umm, github thinks the diffs are outdated, but I think they're not, so my comments should be valid.

This seems to be a github bug.  I've had this happen a few times.
I think it treats any commits made in the individual commit view as outdated.

**Comment by Mailaender** [comment]
Updated.

File: OpenRA.Mods.Common/Traits/SupportPowers/SupportPower.cs [hunk 8]
```diff
@@ -43,11 +43,19 @@ public abstract class SupportPowerInfo : UpgradableTraitInfo
...
-		public readonly string BeaconPoster = null;
+
+		public readonly bool BeaconPaletteIsPlayerPalette = true;
+		[PaletteReference("BeaconPaletteIsPlayerPalette")] public readonly string BeaconPalette = "player";
+
+		public readonly string BeaconImage = "beacon";
+		[SequenceReference("BeaconImage")] public readonly string BeaconPoster = null;
 		[PaletteReference] public readonly string BeaconPosterPalette = "chrome";
+		[SequenceReference("BeaconImage")] public readonly string ClockSequence = "clock";
+		[SequenceReference("BeaconImage")] public readonly string ClockSprite = "clock";
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/SupportPowers/SupportPower.cs]: This line still needs to be removed (there already is ClockSequence above).

File: OpenRA.Mods.Common/Scripting/Global/BeaconGlobal.cs [hunk 9]
```diff
@@ -27,9 +27,10 @@ public BeaconGlobal(ScriptContext context) : base(context)
 
 		[Desc("Creates a new beacon that stays for the specified time at the specified WPos. " +
 			"Does not remove player set beacons, nor gets removed by placing them.")]
-		public Beacon New(Player owner, WPos position, int duration = 30 * 25, bool showRadarPings = true, string palettePrefix = "player")
+		public Beacon New(Player owner, WPos position, int duration = 30 * 25, bool showRadarPings = true)
 		{
-			var playerBeacon = new Beacon(owner, position, duration, palettePrefix);
+			var beacon = owner.PlayerActor.Info.TraitInfo<PlaceBeaconInfo>();
+			var playerBeacon = new Beacon(owner, position, duration, beacon.Palette, beacon.IsPlayerPalette, beacon.BeaconImage, beacon.ArrowSprite, beacon.CircleSprite);
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Scripting/Global/BeaconGlobal.cs]: You forgot to adapt this to the *Sequence renamings.

**Comment by Mailaender** [comment]
Fixed those rebase problems over and over again. Hopefully for the last time.

File: OpenRA.Mods.Common/Traits/SupportPowers/AirstrikePower.cs [hunk 10]
```diff
@@ -169,9 +169,14 @@ public void SendAirstrike(Actor self, WPos target, bool randomize = true, int at
...
 						self.Owner,
 						target - new WVec(0, 0, altitude),
-						Info.BeaconPalettePrefix,
+						Info.BeaconPaletteIsPlayerPalette,
+						Info.BeaconPalette,
+						Info.BeaconImage,
 						Info.BeaconPoster,
 						Info.BeaconPosterPalette,
+						Info.CircleSequence,
+						Info.ArrowSequence,
```
Comment (obrakmann) [inline OpenRA.Mods.Common/Traits/SupportPowers/AirstrikePower.cs]: You've got the order of the arguments flipped around here: `Info.ArrowSequence` comes before `Info.CircleSequence`.

File: OpenRA.Mods.Common/Traits/SupportPowers/NukePower.cs [hunk 11]
```diff
@@ -134,9 +134,14 @@ public override void Activate(Actor self, Order order, SupportPowerManager manag
...
 					order.Player,
 					targetPosition,
-					Info.BeaconPalettePrefix,
+					Info.BeaconPaletteIsPlayerPalette,
+					Info.BeaconPalette,
+					Info.BeaconImage,
 					Info.BeaconPoster,
 					Info.BeaconPosterPalette,
+					Info.CircleSequence,
+					Info.ArrowSequence,
```
Comment (obrakmann) [inline OpenRA.Mods.Common/Traits/SupportPowers/NukePower.cs]: You've got the order of the arguments flipped around here: `Info.ArrowSequence` comes before `Info.CircleSequence`.

File: OpenRA.Mods.RA/Traits/SupportPowers/ParatroopersPower.cs [hunk 12]
```diff
@@ -204,9 +204,14 @@ public Actor[] SendParatroopers(Actor self, WPos target, bool randomize = true,
...
 						self.Owner,
 						target - new WVec(0, 0, altitude),
-						Info.BeaconPalettePrefix,
+						Info.BeaconPaletteIsPlayerPalette,
+						Info.BeaconPalette,
+						Info.BeaconImage,
 						Info.BeaconPoster,
 						Info.BeaconPosterPalette,
+						Info.CircleSequence,
+						Info.ArrowSequence,
```
Comment (obrakmann) [inline OpenRA.Mods.RA/Traits/SupportPowers/ParatroopersPower.cs]: You've got the order of the arguments flipped around here: `Info.ArrowSequence` comes before `Info.CircleSequence`.

**Comment by obrakmann** [comment]
:+1: after fixing the order of the arguments.

**Comment by Mailaender** [comment]
Fixed.

**Comment by obrakmann** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/3094568fe01ea99609fbc035fa81501cc6a0051f)

**Pull request merged**